org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -666,9 +777,10 @@ protected void processAnnotations(CompilationParticipantResult[] results) {
 		results[i].reset(foundAnnotations ? this.filesWithAnnotations.get(results[i].sourceFile) : null);
 	}
 
+	boolean isEcjUsed = Compiler.class.equals(this.compiler.getClass());
 	// even if no files have annotations, must still tell every annotation processor in case the file used to have them
 	for (CompilationParticipant participant : this.javaBuilder.participants)
-		if (participant.isAnnotationProcessor())
+		if (isEcjUsed && participant.isAnnotationProcessor())
```

Comment:
I believe the check for isEcjUsed  could be moved up prior to looping through the participants